app/deliveryoptions/delivery_options.py

```diff
@@ -0,0 +1,318 @@
+
+    for s in surrogates:
+        if s["xReferenceURL"]:
+            surrogate_list.append(s["xReferenceURL"])
+
+    return surrogate_list
+
+
+def generic_builder(
+    delivery_option_data: Union[List, str],
+    record_data: Record,
+    api_surrogate_data: List = None,
+    builder_type: str = "default",
+) -> Union[str, List]:
+    """
+    A generic builder function to handle various delivery option content types.
+
+    Args:
+        delivery_option_data: The data to process
+        record_data: The record object
+        api_surrogate_data: List of surrogate data
+        builder_type: Type of builder to determine special processing
+
+    Returns:
+        Processed content (str or List) based on the builder type
+    """
+    # Handle special case for distressing content
+    dcs_flag = False
+    if builder_type == "description" and has_distressing_content_match(
+        record_data.reference_number
+    ):
+        dcs_flag = True
+
+    # Default HTML building
+    if builder_type == "orderbuttons" and isinstance(
+        delivery_option_data, list
+    ):
+        result = []
+        for item in delivery_option_data:
+            processed_item = {}
+            for key, value in item.items():
+                if key in ["href", "text"]:
+                    processed_item[key] = html_builder(
+                        value,
+                        record_data,
+                        api_surrogate_data=api_surrogate_data,
+                    )
+                else:
+                    processed_item[key] = value
+            result.append(processed_item)
+        return result
+
+    # Standard HTML building
+    return html_builder(
+        delivery_option_data,
+        record_data,
+        api_surrogate_data=api_surrogate_data,
+        dcs=dcs_flag,
+    )
+
+
+def construct_delivery_options(
+    api_result: List, record: Record, request: HttpRequest
+) -> Dict[str, Any]:
+    """
+    Construct delivery options based on record and request information.
+
+    This is the main function called from records.py to build the delivery options
+    for a record.
+
+    Args:
+        api_result: List of delivery options
+        record: The record object
+        request: The HTTP request
+
+    Returns:
+        The constructed delivery options
+    """
+
+    if api_length := len(api_result) > 1:
+        raise ValueError(
+            f"Too many results ({api_length}) from DORIS database for IAID {record.iaid}"
+        )
+
+    delivery_options_context_dict = {}
+
+    reader_type = get_reader_type(request)
+
+    delivery_options_context_dict["reader_type"] = reader_type
+
+    do_dict = read_delivery_options(settings.DELIVERY_OPTIONS_CONFIG)
+
+    # Surrogate links is always present as a list, which can be empty
+    do_surrogate = surrogate_link_builder(api_result[0]["surrogateLinks"])
+
+    if (
+        api_result[0]["options"]
+        == AvailabilityCondition.ClosedRetainedDeptKnown
+    ):
```

Comment:
```suggestion
   api_result_options = api_result[0]["options"]
    if (
        api_result_options
        == AvailabilityCondition.ClosedRetainedDeptKnown
    ):
```
more clearer to have the result in variable as its used in multiple places.

Reply:
> I understand what you are saying but there are three places where api_result[0]["options"] is used and one of them is an assignment to it. Changing it as you suggest would still need the assignment to be retained:
api_result[0]["options"] = AvailabilityCondition.ClosedRetainedDeptUnKnown

I think changing the value of the api results from the original value could be avoided. How does this logic feel?
as in 

```
api_result_options = api_result[0]["options"]
if (
        api_result_options
        == AvailabilityCondition.ClosedRetainedDeptKnown
    ):
        # Special case. Sometimes, for record type 14 (ClosedRetainedDeptKnown), the department name does not match
        # any entry in the DEPARTMENT_DETAILS dictionary. This shouldn't happen but it does. Therefore, reset the type
        # with that for AvailabilityCondition.ClosedRetainedDeptUnKnown
        if not get_dept(record.reference_number, "deptname"):
            delivery_option = do_dict["deliveryOptions"]["option"][AvailabilityCondition.ClosedRetainedDeptUnKnown]
else:
        # Get the specific delivery option for this artefact
        delivery_option = do_dict["deliveryOptions"]["option"][api_result_options]
```

Reply:
You might be right but this is messy and I don't think it adds enough to make it worthwhile. I'm uncomfortable with the code you have suggested as I'm not convinced it's correct. If it is, then I still don't like it!

Reply:
I've made a partial compromise by moving the department check into the main if statement

Reply:
OK, I have changed it again, but possibly simpler than either solution.

Reply:
Suggested to use name api_result_options->availability_condition if you're removing else: part